Book information: refactored and additional features

poire-z opened the pull request:
See discussion in #2986 and individual commit detailed messages:
- TextViewer widget, used when onHold on truncated item in KeyValuePage
- Store book metadata and nb of pages in new settings
- Book information: refactored and additional features

(not sure why github displays `\x`C2`\x`A0 in red, they work and I had added a lot in wikipedia.lua, where they are displayed in red too)

Frenzie commented on frontend/util.lua:
```diff
@@ -400,4 +400,102 @@ function util.splitToArray(str, splitter, capture_empty_entity)
+    end
+end
+
+local HTML_ENTITIES_TO_UTF8 = {
```
This feels oddly familiar but I suppose not from this project.

Frenzie commented on frontend/ui/widget/keyvaluepage.lua:
```diff
@@ -150,6 +153,15 @@ function KeyValueItem:init()
             end
+            -- allow for displaying the non-truncated texts with Hold
+            if Device:isTouchDevice() then
+                self.ges_events.Hold = {
```
Is `registerTouchZones` available here?

poire-z replied:
Looks like none in frontent/ui/widget/ uses it, so I assume it's only things in frontend/apps/ that should.

Frenzie replied:
Pinging @houqp who implemented the system. Obviously it's most relevant on documents, but since most of the widgets predate it that doesn't necessarily mean it wasn't intended to be used there in the future.

Frenzie commented on frontend/ui/widget/textviewer.lua:
```diff
@@ -0,0 +1,238 @@
+local _ = require("gettext")
+local Screen = Device.screen
+
+--[[
```
I'm fairly sure that ldoc only picks it up if you stick this at the top of the file, although for such a paltry bit of documentation I suppose it doesn't matter. :-P

Frenzie commented:
Looks okay. Could you add a screenshot of TextViewer to save me some time?

poire-z commented:
Yes.

<kbd>![bookinfo](https://user-images.githubusercontent.com/24273478/27761596-7ccecc14-5e5f-11e7-812d-a29cf78d9fe1.png)</kbd>

<kbd>![textviewer](https://user-images.githubusercontent.com/24273478/27761600-84a168a2-5e5f-11e7-8343-111fa5411320.png)</kbd>

Hzj-jie commented on frontend/apps/filemanager/filemanagerbookinfo.lua:
```diff
@@ -0,0 +1,186 @@
+    local file_mode = lfs.attributes(file, "mode")
+    if file_mode ~= "file" then
+        return false
+    end
```
Seems you forgot to return true. And this function can be shortened.
P.s.the original logic in file manager requires not is_pic.

poire-z replied:
Indeed, I forgot to return true :| The strange thing is that it still works like I expected it to work, and it shouldn't now that you pointed that to me :) I'll investigate :)
Someone indeed used `is_pic` in the original code, because calling `document:getProps()` on a `picdocument` just throw an error (and make the reader crash if you don't catch it).
As it's a bug in picdocument (ffi/C part) and should probably be fixed in another PR, I decided there and elsewhere to use ``pcall(document.getProps, document)`` to catch the case when it's a picdocument, so when pidocument is fixed, it won't need a change here. Also, picdocument could be made to return JPEG images's comment as Description, so we can save/display it.

poire-z replied:
OK, turns out that if you set in one of the buttons dict `{ enabled = nil }`, that `enabled` key just does not exist in the table. So, the default `enabled = true` in the Button class definition is used. That's why it worked. But i'll shorten it and have it return real true and false.

Hzj-jie commented on frontend/apps/reader/readerui.lua:
```diff
@@ -342,6 +349,16 @@ function ReaderUI:init()
+    if not ok then
+        doc_props = {}
+    end
+    self.doc_settings:saveSetting("doc_props", doc_props)
```
Would you mind to store these non-changeable settings into a different file? Include fastDigest(). The docsettings needs to be optimized.

poire-z replied:
I would :| It would need us to manage another read/save/flush setting.
It's already there in the current settings, under the ``stats`` key. I'm just easing the transition when statistics plugin won't update this stats key anymore. There's one single file today, I don't want to be the one introducing a second file :) (also see note above about `DocSettings:saveSetting()` only updating a key, and not flushing to disk).
(My aim is to add features, while keeping code clean and neat, as it is already. Not so much about a big refactoring. I'm quite often just cut and pasting from other existing code that works :)

Hzj-jie replied:
If you do not want to work on splitting the configuration file, I am OK with this. The configuration file is badly less optimized. We have to update it. https://github.com/koreader/koreader/issues/2789

Hzj-jie commented on frontend/ui/widget/textviewer.lua:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    local buttons = {
+        {
```
Nit: I would suggest not to newline between several {.

poire-z replied:
Again, I followed what is done in most other existing code I took inspiration from, sticking to tradition :)

Frenzie replied:
I'd suggest in favor for clarity. Why no newline?

Hzj-jie commented on frontend/ui/widget/textviewer.lua:
```diff
@@ -0,0 +1,238 @@
+    end)
+end
+
+function TextViewer:onCloseWidget()
```
I remember setDirty logic is embedded in UIManager when a widget is shown or closed. Would you mind to have a try?

poire-z replied:
I tried it, and it indeed works as well without it.
This `:onCloseWidget()` is in 30% of the widgets, so it came in mine from the widget I choose as a base for this new one (dictquicklookup).
I can remove it, unless I miss a non-obvious need for it that would explain why it is there in some existing widgets.

Hzj-jie commented on frontend/util.lua:
```diff
@@ -400,4 +400,102 @@ function util.splitToArray(str, splitter, capture_empty_entity)
+    ["&quot;"] = '"',
+    ["&apos;"] = "'",
+    ["&nbsp;"] = "\xC2\xA0",
+    ["&#(%d+);"] = function(x) return util.unicodeCodepointToUtf8( tonumber(x) ) end,
```
The space before and after tonumber() do not match the common coding style, and also the following line.

poire-z replied:
Will fix that.

Hzj-jie commented on frontend/util.lua:
```diff
@@ -400,4 +400,102 @@ function util.splitToArray(str, splitter, capture_empty_entity)
+    ["&apos;"] = "'",
+    ["&nbsp;"] = "\xC2\xA0",
+    ["&#(%d+);"] = function(x) return util.unicodeCodepointToUtf8( tonumber(x) ) end,
+    ["&#x(%x+);"] = function(x) return util.unicodeCodepointToUtf8( tonumber(x,16) ) end,
```
A more comprehensive list can be found at https://stackoverflow.com/questions/14899734/unescape-numeric-xml-entities-with-lua.
p.s. You may need to put &amp; at the end of the list, otherwise '&amp;quot;' will be converted to '"', but indeed it should be '&quot;'.

Why do not name this function "htmlDecode"?

According to the following usage, I would suggest to return a tuple from this function: `{string decoded, bool changed}`. i.e.
```
function util.htmlDecode(text)
  if text contains html entities then
    return decode(text), true
  else
    return text, false
  end
end function
```
then a `util.htmlDecodeIfEncoded(text)` can be written as,
```
function util.htmlDecodeIfEncoded(text)
  if text contains html tags then
    return text, false
  else
    local decoded, result = util.htmlDecode(text)
    if result then
      return decoded, true
    else
      return text, false
    end
  end
end function
```
So you won't need to handle the double encoded html in `util.htmlToPlainTextIfHtml()`.

poire-z replied:
OK, I must say I'm not that proud about this util stuff. I just did enough so that the descriptions in the dozen of epub with descriptions I have would ALL display nice.
But I guess if we put that in util, it should be complete and not just enough for descriptions of a dozen books :|

I'll have a go with following your suggestions and pointers.
Your stackoverflow link lists many entities, most indeed used in french. In my dozen french book, I did not find a single such named entities (like ``&eacute;``). I don't know if I should double the size of our ``util.lua`` with these, or just add comments about this function limitations.

Also, this comprehensive list lists mostly latin named entities. Dunno if there are other ones in other alphabets/languages (I think there should be, greek probably have ``&alpha;``, ``&beta;`` ...).
Should I add such a big list of named entities in util.lua?

poire-z replied:
Well, I tried to roll things up as you suggested, but it got quite confusing, and I couldn't make it as readable as I think it was.
So, I'd like to keep it as it is:
- a simple htmlToPlainText(text) that can be used as is
- a htmlToPlainTextIfHtml(text) that does some cheap detection of HTML according to 2 simple rules :
  - unbalanced < and >
  - encoded HTML tags `"&lt;%a+&gt;"` directly (no need to decode the whole text and check for tags, cheaper to check for encoded tags directly - and i dont want to decode it if there's no tag in it).

To me, it looks more readable, and explicitely describes my flow of thiking.

I added some comments about it being limited and for use with "simple" HTML. When we get wilder input, we may rework it.

Hzj-jie commented on frontend/util.lua:
```diff
@@ -400,4 +400,102 @@ function util.splitToArray(str, splitter, capture_empty_entity)
+    return text
+end
+
+--- Convert HTML to plain text
```
This comment and the function name are a little bit strange, indeed it handles only <br> and <p>, but remove all other tags. I think removeHtmlTags() may be a better name.

poire-z replied:
But I don't want to _remove Html Tags_ :)
About decode vs convert in your previous comment, I don't consider HTML as some ``encoding`` that needs to be `decoded`.
I consider it `` formatting``. And in this case, I want to convert it to another formatting (plain text), and the only HTML formatting tags that make sense in plain text are the newlines (to keep paragraphs, see the description screenshot). That's why `<br>` and `<p>` are the only ones that I handle (and if this function were needed to be super good, it should also handle ``<div>`` and ``<span style="display: block">``, and first remove HTML comments, styles and scripts, and it would be tough and long to implement that correctly).
I need it to work on simple HTML like I have found in epub's metadata ``<dc:description>``. If people were to open issues with samble books with more complex ``<dc:description>``, we may need to revisit that.

poire-z replied:
OK, just seeing I forgot to update this comment:
`--- Remove HTML tags from text if text seems to be HTML
`
which I should have rewritten, like I did for the previous function's comment, to:
`--- Convert HTML to plain text if text seems to be HTML
`
(when I started, i actually was _removing HTML tags_ :) before I realized i needed to do conversion, and so the names and comments became about _converting HTML to plain text_, except this comment that I forgot, which may have led you on the _remove html_ track - sorry)

Hzj-jie commented on frontend/util.lua:
```diff
@@ -400,4 +400,102 @@ function util.splitToArray(str, splitter, capture_empty_entity)
+    -- if it is HTML, we should get the same number of '<' and '>'
+    -- that surround tags
+    local _, nb_gt, nb_lt
+    _, nb_gt = text:gsub(">", "")
```
I think you can match "<\w+.*>" instead of ">" and "<". The later two characters are commonly used in JS.

poire-z replied:
Yup, will do that.

poire-z replied:
Had a try with ``"<%w+.->"``, and both detect correctly if HTML or not on my sample books.

But given the following fictionnal non-html:    ``if (a<b and a>c)``, both our solutions will detect it as HTML, and it will be a false positive.
Given the following non-html:    ``if (a<b and a>c and d>e)``, detecting a tag will find one, and decide it is HTML. Counting ``<`` and ``>`` like I originally did will also find them, but it will detect the numbers are different and that we have unbalanced ``<`` and ``>`` , and decide it is not HTML.
So, as quick and dirty ways (which are enough for our uses) to decide if HTML or not, i'll stick with the one with one less false positives chance.

Hzj-jie replied:
I think counting < and > will introduce more true negative instead of concerning false positive: Html contents contain js will be treated as non-html.
But your example is not accurate, contents contain js are usually Html.

poire-z replied:
OK, as both work on my samples, I don't mind.
Rebased with  `"<%w+.->"` .

Hzj-jie commented on frontend/util.lua:
```diff
@@ -400,4 +400,102 @@ function util.splitToArray(str, splitter, capture_empty_entity)
+    else
+        -- no <tag> found
+        -- but we may meet some text badly twicely encoded html containing "&lt;br&gt;"
+        local nb_encoded_tags
```
Typically decoding html for several times until a tag is found won't break the HTML. So maybe you can put this logic at the beginning of the function.

poire-z replied:
Unless it's technical HTML talking about HTML and having litteral `&amp;` like this comment does :)
(which we should hardly ever meet in ``<dc:description>``).

poire-z commented on frontend/apps/reader/modules/readerfooter.lua:
```diff
@@ -546,6 +546,7 @@ end
 function ReaderFooter:onPageUpdate(pageno)
     self.pageno = pageno
     self.pages = self.view.document:getPageCount()
+    self.ui.doc_settings:saveSetting("doc_pages", self.pages) -- for Book information
```
We dont't need to save the page count every time page turns.
But note that:
```lua
    --- Saves a setting.
    function DocSettings:saveSetting(key, value)
        self.data[key] = value
    end
```

so `saveSetting` just updates a key, it does not save the setting to disk (``DocSettings:flush()`` does that).
What I need is to have, in the sidecar file, the number of pages of the book. This number of pages is fixed for PDF, but it may change for crengine docs, and it does when you change fonts, switch orientation... And I want it updated in the settings data structure when that happen, so it is saved to the sidecar file when one leaves Reader.
Logically, we may find out all the function that may have the number of pages changed, and add ``self.ui.doc_settings:saveSetting("doc_pages", self.view.document:getPageCount())`` ti each of them, but we may miss a few.
But when any of these happen (font, orientation, hyphenation, floating punctuation), `PageUpdate` event is send, and all `onPageUpdate` are called (it's also called when you just turn page, as you said).
And in `readerfoot.lua` , we have:
```lua
function ReaderFooter:onPageUpdate(pageno)
    self.pageno = pageno
    self.pages = self.view.document:getPageCount()
    self:updateFooterPage()
end
```

and it is actually the only ``onPageUpdate`` that is already calling ``self.view.document:getPageCount()``, so, by putting what I put there, I put it just once, and I avoid the cost of doing more calls to  ``document:getPageCount()``.
I could also put it more logically in both readerrolling and readerpaging, or in the single readerview, for the cost of two or one additionnal ``document:getPageCount()``.

(I don't get what fastDigest() does, and how it could help with what I want to do.)

Hzj-jie replied:
I misunderstood the page count, if it changes time to time, my comment is not proper.

poire-z commented on frontend/apps/filemanager/filemanagerbookinfo.lua:
```diff
@@ -0,0 +1,186 @@
+--[[--
```
This PR already makes it available in Reader (I mentionned that in the commit message).
`filemanagerbookinfo.lua` is similar in many aspects to `apps/filemanager/filemanagerhistory.lua`: it is imported in Reader, uses `registerToMainMenu`/`addToMainMenu` to add itself to Reader's menu.
I think it belongs into `frontend/apps`, it just get data from files, settings, make some data structure out of them (_business_ logic) and call a real generic widget: KeyValuePage (_view_ logic). Just like filemanagerhistory.lua calls a the generic Menu widget.
Putting it into `frontend/ui/widget` would need this code to be splitted into 2 files (one in apps/ for the menu, one in ui/widget/), and I don't see where we would use again the ui/widget part (which would contain some business logic).

poire-z commented:
OK, rebased as I'd like to keep these 3 distinctive ommits.
Changes:
frontend/ui/widget/textviewer.lua: removed unnecessary`TextViewer:onCloseWidget()`

frontend/apps/filemanager/filemanagerbookinfo.lua: shortened `BookInfo:isSupported(file)` , which returns real true/false

frontend/util.lua: removed spaces surrounding `tonumber()`s , moved `["&amp;"]` at end of list, added comments to `util.htmlToPlainText()` and `util.htmlToPlainTextIfHtml()` about their limitations and usage with _simple_ HTML.
I didn't add a big list of named entities, as I didn't see any in my sample books, and I didn't want to discriminate between alphabets and favour latin :)

Frenzie commented:
There's a list here btw https://dev.w3.org/html5/html-author/charref

poire-z commented:
Want me to add them to a file and use them ?
I don't mind parsing this page and make a lua table out of them. May be not in frontend/util.lua cause I'd like to keep it small, but in frontend/htmlentities.lua (?) , which would be imported only in the util function that would need them ?

poire-z commented:
> I remember setDirty logic is embedded in UIManager when a widget is shown or closed.

Had to put back `TextViewer:onCloseWidget()`. Screen was not refreshed/redrawn when I closed a TextViewer launched from a filemanager menu item (tested by adding one item to filemanagermenu.lua 's Developer's option to display settings.reader.lua with TextViewer).

poire-z commented:
Added a 4th commit for easier review, i'll squash it into the 3rd before merging.

Is there anything still preventing this being merged ? (I would need some of its stuff to add description/keyword support to CoverBrowser plugin db.)

Frenzie commented:
I haven't been able to look at it as closely as I'd like but I'd say that nothing jumping out is always good. ;-)

poire-z commented:
Thank you very much :)
(I'll add a commit to coverbrowser plugin tomorrow, once a new nightly has been generated, so that curious people don't crash in the meantime)
